batcher.go

```diff
@@ -0,0 +1,270 @@
+package batcher
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"golang.org/x/time/rate"
+)
+
+const (
+	DefaultMaxBatchSize = 100
+	DefaultMinBatchSize = 1
+	DefaultMaxHandlers  = 1
+	DefaultTicker       = 17 * time.Millisecond
+)
+
+type batcherType int
+
+const (
+	MainBatcher batcherType = iota
+	TimerBatcher
+)
+
+type Result[V any] struct {
+	Value V
+	Err   error
+}
+
+// New create new batcher
+func New[K comparable, V any](handler func(ctx context.Context, keys []K) []*Result[V], opts ...Option[K, V]) *batcher[K, V] {
+	o := &opt[K, V]{
+		MaxBatchSize: DefaultMaxBatchSize,
+		MinBatchSize: DefaultMinBatchSize,
+		MaxHandlers:  DefaultMaxHandlers,
+		Ticker:       DefaultTicker,
+	}
+
+	for _, oo := range opts {
+		oo(o)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	hwrap := wrapHandler(o.Cache, o.Timeout, handler)
+
+	b := &batcher[K, V]{
+		ctx:     ctx,
+		opt:     o,
+		handler: hwrap.Handle,
+		pending: make([]waiter[K, V], 0, o.MaxBatchSize),
+		cancel:  cancel,
+	}
+	go b.batchTicker()
```

Comment:
This is to prevent more waiting times. For example, you set that the minimum batch size is 10. You may get a situation where 9 keys are waiting, and the 10th one will come after 1 second. Then we get that all 9 records will just wait 1 second. It's not very good